Feature/multi qubit time domain

stephlazar opened the pull request:
Multi-qubit time domain measurements.

Changes proposed in this pull request:
- obviously, the n_qubit time domain measure functions + sequences
- fixes a small bug in analysis_toolbox.py get_timestamps_in_range which would fail if the day folder being searched contained any file in addition to just folders, or if it contained a folder whose first 6 characters were not of the format HHMMSS.
- cleans up the MultiQubit_TimeDomain_Analysis. Splits some long methods like process_data into sub-methods. Should improve readability. 
- improves the get_indices method or CalibrationPoints. This method would return the wrong indices if we used preselection or active_reset. I now added the class attribute prep_params (which it gets from the create_segments method) which will be used by get_indices.
- adds get_multiplexed_readout_detector_function to multi_qubit_module.py. This is a nice wrapper around get_multiplexed_readout_detector_functions.

FYI @christiankraglundandersen @jo3141 

This branch has been tested at XLD. You can see the 7 qubit time domain measurements on 20200512.

Commit:
Prepares branch for pull request.

Adds multi-qubit time domain sequences to multi_qubit_tek_seq_elts.py and
 measure functions to multi_qubit_module.py.

Cleans up MultiQubit_TimeDomain_Analysis.
Fixes bugs in QScaleAnalysis.

Adds get_meas_obj_value_names_map.

Fixes bug in get_timestamps_in_range which would fail if the day folder contained
  files in addition to folders or if the folder name did not have the format HHMMSSS_X.

Improves CalibrationPoints to get_indices based on preparation parameters.

Commit:
Sets upload=False in calls to sequence functions in all n_qubit time domain measure functions.

Commit:
Improves handling of sweep points in QScaleAnalysis. Now The analysis can produce the raw data plots even if
 the sweep points are saved as unique values in the metadata.

Commit:
Merge branch 'qudev_master' into feature/multi_qubit_time_domain

Commit:
Merge branch 'qudev_master' into feature/multi_qubit_time_domain

Commit:
Merge branch 'qudev_master' into feature/multi_qubit_time_domain

Commit:
Merge branch 'qudev_master' into feature/multi_qubit_time_domain

antsr commented on pycqed/measurement/calibration_points.py:
```diff
@@ -72,7 +75,7 @@ def get_states(self, qb_names=None):
                 for qbn in qb_names}
 
-    def get_indices(self, qb_names=None):
+    def get_indices(self, qb_names=None, prep_params=None):
```
a brief comment would be useful, e.g. 
> prep_params as returned by `QuDev_Transmon.prep_params`

stephlazar replied:
Brief and not-so-brief comments are always useful :p 
I added it now, thanks :)

Commit:
Adds prep_params to docstring of get_indices.

chellings commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -155,6 +155,29 @@ def get_operation_dict(qubits):
     return operation_dict
 
 
+def get_multiplexed_readout_detector_function(qubits, det_type, **kw):
```
Due to the similar names, it is difficult to tell the difference between get_multiplexed_readout_detector_function and get_multiplexed_readout_detector_functions. Docstrings would help, but maybe it could also be checked whether it makes sense to either a) combine them in one function if they nearly do the same thing, or b) change the name of the new function that that the difference gets clear.

stephlazar replied:
I renamed it now to get_n_qubit_detector_function. This is a wrapper around get_multiplexed_readout_detector_functions. I would keep it distinct from the latter since I wouldn't want the default values for nr_averages and nr_shots to be taken from the qubits. But maybe it makes sense? I'm not sure. What do you think, should we combine the two functions?

chellings commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -155,6 +155,29 @@ def get_operation_dict(qubits):
 
 
+def get_multiplexed_readout_detector_function(qubits, det_type, **kw):
+    # only used by classifier detector
```
Not clear whether this comment is supposed to apply to the whole function or only to particular lines

stephlazar replied:
I now added docstring so it should be clear.

chellings commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -3190,4 +3213,558 @@ def measure_ro_dynamic_phases(pulsed_qubit, measured_qubits,
+    elif len(qubit_names) > 5:
+        msmt_suffix = '_{}qubits'.format(len(qubit_names))
+    else:
+        msmt_suffix = '_qbs{}'.format(''.join([i[-1] for i in qubit_names]))
```
'_'.join([i.strip('qb') for i in qubit_names]) # for S17 compatibility :-)

stephlazar replied:
Good point! :) Thanks!

chellings commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -3190,4 +3213,558 @@ def measure_ro_dynamic_phases(pulsed_qubit, measured_qubits,
+    tda.MultiQubit_TimeDomain_Analysis(qb_names=[qb.name for qb in measured_qubits])
+
+
+def get_multi_qubit_msmt_suffix(qubits):
```
docstring missing

stephlazar replied:
Added

Commit:
Merge branch 'qudev_master' into feature/multi_qubit_time_domain

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -3819,8 +3842,19 @@ def process_data(self):
+            unique_sp = np.unique(sweep_points[:3])
+            if unique_sp.size > 1:
+                sweep_points = np.repeat(sweep_points, 3)
+            # replace in proc_data_dict; otherwise pltting in base class fails
```
pltting :-)
apart from that, not sure about these changes, but since the code is tested, it should be fine :-)

stephlazar replied:
Yes, I tested the code.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -3933,9 +3967,10 @@ def prepare_plots(self):
+                    xlabel = self.raw_data_dict['sweep_parameter_names']
+                    xunit = self.raw_data_dict['sweep_parameter_units']
+                if np.ndim(xunit) > 0:
+                    xunit = xunit[0]
```
what about xlabel? Is it intended that xlabel might be a list?

stephlazar replied:
I am not entirely sure, so I added the same check for the xlabel as well :p

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -291,6 +313,24 @@ def process_data(self):
+
+        self.create_sweep_points_dict()
+        self.create_meas_results_per_qb()
+
```
New structure looks good! (Haven't checked every detail of it.)

stephlazar replied:
👍

chellings reviewed (approved):
Thank you for this nice new features. I have seen the test measurements on XLD and couldn't resist playing with some further examples when testing on XLD after the software update. Looks all very nice. I just added a few comments that I had while looking at the code.

Commit:
Renames get_multiplexed_readout_detector_function to get_n_qubit_detector_function.
Adds docstring to get_n_qubit_detector_function and get_multi_qubit_msmt_suffix.
Adds check for whether xlabel is a list in QScaleAnalysis.prepare_plots().

Commit:
Merges get_n_qubit_detector_function into get_multiplexed_readout_detector_functions.
Adds get_correlation_channels.
Modifies get_multi_qubit_msmt_suffix to concatenate full qubit names for < 5 qubit measurements.

nathlacroix commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -249,12 +236,46 @@ def process_data(self):
                 {qbn: {'sweep_points': list(hard_sweep_params.values())[0][
                     'values']} for qbn in self.qb_names}
+        elif self.sp is not None:
+            self.mospm = self.get_param_value('meas_obj_sweep_points_map')
```
Do we get information somewhere about what this mospm has to look like? Since I've never seen it before, I looked for it in doc strings but couldn't find much related to it (maybe I just didn't search at the right place)

stephlazar replied:
Yes, this is parameter can be obtained by calling the method get_meas_obj_sweep_points_map in SweepPoints, and there is a docstring there.

nathlacroix commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -3819,8 +3842,19 @@ def process_data(self):
-                'msmt_sweep_points']
+            sweep_points = deepcopy(self.proc_data_dict['sweep_points_dict'][
+                                        qbn]['msmt_sweep_points'])
+            unique_sp = np.unique(sweep_points[:3])
```
not sure I understand this magic here :D why do you need to cut after 3 sweep points? having a magic number here seems a bit odd to me at first sight

stephlazar replied:
For the qscale measurement, each sweep point needs to be repeated 3 times. So what I do here is I check if the sweep points are already repeated 3 times by looking if the first 3 entries are identical (as they would need to be) or not. The issue is that in the single qubit routine we already save the repeated sweep points in metadata while in the new multi-qubit routine we do not (this is because of the new SweepPoints object).

nathlacroix replied:
ok, I see. Sorry I didn't see this was in the qscale analysis (the code was folded and I thought it was the baseanalysis). Maybe then a small comment saying that it is for backwards compatibility with single qb routine and than it can be removed once the latter is re-written with the new Sweeppoint objects would be useful. Something like : `# Verifies whether sweep points are stored with repetition or not. For backwards compatibility with QudevTransmon.measure_qscale() that does not (yet) use Sweeppoints object.`

stephlazar replied:
I already added a comment in that direction. See lines 3847-3850 right above this line.

nathlacroix replied:
I think my main point was to show why we needed to check i.e. the backwards compatibility so that once we change it in measure_qscale we can remove this check. As this was just a really minor thing I committed myself, hope that's ok for you.

nathlacroix commented on pycqed/measurement/calibration_points.py:
```diff
@@ -19,10 +19,12 @@ def __init__(self, qb_names, states, **kwargs):
         self.states = states
```
Thanks for fixing this issue :) I just have a few questions/comments on the implementation (see comments below)

nathlacroix commented on pycqed/measurement/calibration_points.py:
```diff
@@ -19,10 +19,12 @@ def __init__(self, qb_names, states, **kwargs):
         self.states = states
         default_map = dict(g=['I '], e=["X180 "], f=['X180 ', "X180_ef "])
         self.pulse_label_map = kwargs.get("pulse_label_map", default_map)
+        self.prep_params = None
```
Is there a specific reason why you have the prep params as a state for the cal_point object? In principle, a cal point could remain independent of the preparation and could be used with various preparations, right? We could in that case have the prep_params as an argument of the get_indices() function, which, as you pointed out, needs to know which preparation was made. What do you think? 
One reason for which I imagine you could have done it in that way is to save the prep_params as part of the cal point object (?). Ideally, but I guess we should be able to retrieve them from the hdf5 file in any case.

nathlacroix commented on pycqed/measurement/calibration_points.py:
```diff
@@ -19,10 +19,12 @@ def __init__(self, qb_names, states, **kwargs):
     def create_segments(self, operation_dict, pulse_modifs=dict(),
                         **prep_params):
         segments = []
+        self.prep_params = prep_params
```
same comment as above, I'm not sure I understand why we should save the prep_params as part of the state of the calibration_point. to me, those are two different things and the prep_params are just needed to create the segments and get the right indices, it could end up being odd if that changes the cal_point_object depending on the preparation parameters when creating segments, no?

nathlacroix commented on pycqed/measurement/calibration_points.py:
```diff
@@ -72,26 +75,44 @@ def get_states(self, qb_names=None):
         """
+        if prep_params is None:
+            prep_params = self.prep_params
+        if prep_params is None:
+            prep_params = {}
+        prep_type = prep_params.get('preparation_type', 'wait')
```
I think these would be the lines I would keep and I would remove the state in self. But maybe I'm just missing something about the usecase

nathlacroix commented on pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py:
```diff
@@ -2152,3 +2152,375 @@ def ro_dynamic_phase_seq(hard_sweep_dict, qbp_name, qbr_names,
+    len_swp_pts = len(list(sweep_points[0].values())[0][0])
+    final_pulses = np.squeeze(np.reshape(reordered_pulses,
+                                         (3*len_swp_pts, -1))).tolist()
+    # print(len(final_pulses))
```
delete comments

stephlazar replied:
Thanks!

nathlacroix commented on pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py:
```diff
@@ -2152,3 +2152,375 @@ def ro_dynamic_phase_seq(hard_sweep_dict, qbp_name, qbr_names,
+             for p in swept_pulses]
+        final_pulses.append(swept_pulses_with_prep)
+
+    # print()
+    # print(len(final_pulses))
```
delete comments

stephlazar replied:
thanks!

nathlacroix commented on pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py:
```diff
@@ -2152,3 +2152,375 @@ def ro_dynamic_phase_seq(hard_sweep_dict, qbp_name, qbr_names,
+        params_to_sweep = {
+            f'Qscale_*_{qbn}.motzoi': list(sweep_points[0].values())[j][0]
+            for j, qbn in enumerate(qubit_names)}
+        # print(params_to_sweep)
+        # print(len(pulse_list))
```
delete print statements (applies to others also which are commented)

stephlazar replied:
thanks!

nathlacroix commented on pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py:
```diff
@@ -2152,3 +2152,375 @@ def ro_dynamic_phase_seq(hard_sweep_dict, qbp_name, qbr_names,
     return seq, np.arange(seq.n_acq_elements())
+
+
+## Multi-qubit time-domain sequences ##
```
Nice job writing these sequences! 👍 
In the future, I think it would however be nice to have a generic way of expanding a sequence done on one qubit to many qubits in parallel. I guess this is close to what we can currently do on the pulse level with the blocks, but we haven't really thought of this use-case in particular.  Should we create an git issue for this feature? @stephlazar @antsr @chellings  what do you think? Because here, there is about 90% overlap with single qubit sequences, right? If for some reason we have to modify the single qubit sequences, then we'll have to modify this code too, which makes it a bit clumsy. But I think for the short term it make sense to keep it like this.

stephlazar replied:
I think we should always create general n-qubit sequences that work for the case nr_qubits = 1 :P

nathlacroix reviewed (commented):
Very nice new features and several fixes. Thanks!

Commit:
Renames get_sweep_points_map to get_meas_obj_sweep_points_map in SweepPoints.
Removes commented out lines in n_qubit_qscale_seq.

Commit:
Merge branch 'qudev_master' into feature/multi_qubit_time_domain

Commit:
Fixes bug in MultiQubit_TimeDomain_Analysis.process_data() where variable data_filer was unused
 for measurements with active reset.
Removes the attribute prep_params from CalibrationPoints.

stephlazar reviewed (commented):
I now removed the prep_params attribute from CalibrationPoints. Thanks @nathlacroix for the suggestion!

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -300,7 +339,9 @@ def process_data(self):
-            self.cal_states_dict = self.cp.get_indices()[self.qb_names[0]]
+            self.cal_states_dict = self.cp.get_indices(
+                self.qb_names, prep_params)[self.qb_names[0]]
+            print(self.cal_states_dict)
```
delete debug print statement

nathlacroix replied:
print statement still here

Commit:
Merge branch 'qudev_master' into feature/multi_qubit_time_domain

nathlacroix committed:
Clarify comment and remove print statement

nathlacroix reviewed (approved):
Looks good to me, I just committed the removal of a print statement and a minor clarification to a comment.  @stephlazar since you are at the lab today I might do the merge if I see you don't have time to do it by this afternoon. Hope that's fine for you.

stephlazar commented:
@nathlacroix yes please, go ahead with the merge if you have time.

stephlazar commented:
and thanks a lot for taking care of this!
